Cython/Utility/ExtensionTypes.c

```diff
@@ -67,24 +68,35 @@ static int __Pyx_PyType_Ready(PyTypeObject *t) {
         // For details, see https://github.com/cython/cython/issues/3603
         PyObject *ret, *py_status;
         int gc_was_enabled;
-        PyObject *gc = PyImport_Import(PYUNICODE("gc"));
-        if (unlikely(!gc)) return -1;
-        py_status = PyObject_CallMethodObjArgs(gc, PYUNICODE("isenabled"), NULL);
+        static PyObject *gc = NULL;
+        static PyObject *gc_isenabled = NULL, *gc_enable = NULL, *gc_disable = NULL;
+        if (unlikely(!gc)) {
+            gc = PyImport_Import(PYUNICODE("gc"));
+            if (unlikely(!gc)) return -1;
+            gc_isenabled = PyObject_GetAttr(gc, PYUNICODE("isenabled"));
+            gc_enable = PyObject_GetAttr(gc, PYUNICODE("enable"));
+            gc_disable = PyObject_GetAttr(gc, PYUNICODE("disable"));
```

Comment:
Note that it isn't legal in CPython to (potentially) call into Python code or to use arbitrary C-API functions with a live exception set. You have to split the error handling here.